### PR TITLE
Load OMOP CDM dataset during deployment

### PR DESCRIPTION
**Pull request by guybartal**
# PR for issue #97 

## What is being addressed
Load OMOP CDM dataset during deployment

## How is this addressed
- Download and unzip OMOP CSV files from provided storage
- Create database on Azure Database for PostgreSQL for OMOP
- Create CDM schema and tables, primary keys and constrains in the database on Azure Database for PostgreSQL
- Load PostgreSQL CDM tables from CSV files using psql
- Create indexes for the tables in PostgreSQL
- Add CDM source into WebAPI application database

### CODE REVIEW

File: infra/omop_cdm.bicep [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = {
+  name: keyVaultName
+}
+
+// Store OMOM CDM user password in keyvault
+resource postgresAdminSecret 'Microsoft.KeyVault/vaults/secrets@2022-07-01' = {
+  name: postgresOMOPCDMUserSecretName
+  parent: keyVault
+  properties: {
+    value: postgresOMOPCDMpassword
```
Comment (anatbal) [inline infra/omop_cdm.bicep:73]: Should this be here or maybe in the keyvault module? WDYT?

File: infra/omop_cdm.bicep [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  }
+}
+
+// Store the OMOP CDM JDBC connection string in keyvault
+resource postgresOMOPCDMJDBCConnectionStringSecret 'Microsoft.KeyVault/vaults/secrets@2022-07-01' = {
+  name: postgresOMOPCDMJDBCConnectionStringSecretName
```
Comment (anatbal) [inline infra/omop_cdm.bicep:79]: Same as above, for all secrets storing

File: infra/omop_cdm.bicep [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  parent: postgresServer
+  properties: {
+    charset: 'utf8'
+    collation: 'en_US.utf8'
+  }
+}
+
+// Create a managed identity for the deployment script
+resource managedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' = {
+  name: 'managed-identity-omop-cdm'
```
Comment (anatbal) [inline infra/omop_cdm.bicep:98]: Maybe we can use the same identity that was already created for running the previous sql scripts?

File: infra/omop_cdm.bicep [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    supportingScriptUris: [
+      'https://raw.githubusercontent.com/OHDSI/CommonDataModel/main/inst/ddl/5.4/postgresql/OMOPCDM_postgresql_5.4_ddl.sql'
+      'https://raw.githubusercontent.com/OHDSI/CommonDataModel/main/inst/ddl/5.4/postgresql/OMOPCDM_postgresql_5.4_constraints.sql'
+      'https://raw.githubusercontent.com/OHDSI/CommonDataModel/main/inst/ddl/5.4/postgresql/OMOPCDM_postgresql_5.4_primary_keys.sql'
+      'https://raw.githubusercontent.com/OHDSI/CommonDataModel/main/inst/ddl/5.4/postgresql/OMOPCDM_postgresql_5.4_indices.sql'
+      'https://raw.githubusercontent.com/microsoft/OHDSIonAzure/${branchName}/templates/ohdsi-webapi/sql/create_omop_schemas.sql'
+      'https://raw.githubusercontent.com/microsoft/OHDSIonAzure/${branchName}/templates/ohdsi-webapi/sql/add_omop_source.sql'
+    ]
+    cleanupPreference: 'OnSuccess'
+    retentionInterval: 'P1D'
```
Comment (anatbal) [inline infra/omop_cdm.bicep:190]: Should it be PT1H?

File: infra/omop_cdm.bicep [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  kind: 'AzureCLI'
+  identity: {
+    type: 'UserAssigned'
+    userAssignedIdentities: {
+      '${managedIdentity.id}': {}
+    }
+  }
+
+  properties: {
+    forceUpdateTag: utcValue
```
Comment (anatbal) [inline infra/omop_cdm.bicep:114]: Do we want to keep the forceUpdatetag?

File: infra/main.bicep [hunk 6]
```diff
@@ -1,19 +1,45 @@
...
-param suffix string = uniqueString(utc)
+
+@description('Unique string to be used as a suffix for all resources')
+param suffix string = uniqueString(utcNow())
+
+@description('The name of the branch to use for downloading sql scripts')
 param branchName string = 'v2'
 
+@description('The url of the container where the cdm is stored')
+param cdmContainerUrl string
```
Comment (anatbal) [inline infra/main.bicep:19]: you referenced this as secure, should't it be secure here as well?

File: infra/omop_cdm.bicep [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+@description('The name of webapi CDM database')
+param postgresAtlasDatabaseName string
+
+@description('Used to download the sql scripts from the GitHub repository, this should point to the branch you want to use')
+param branchName string
+
+@description('The name of the postgres server')
+param postgresServerName string
+
+@secure()
```
Comment (anatbal) [inline infra/omop_cdm.bicep:17]: Is this supposed to be secure?

File: infra/omop_cdm.bicep [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+@description('The URL of the container where the CDM data is stored')
+param cdmContainerUrl string
+
+@secure()
+@description('The SAS token to access the CDM data')
+param cdmSasToken string
+
+@secure()
+@description('The name of the OMOP CDM database')
+param postgresOMOPCDMDatabaseName string
```
Comment (anatbal) [inline infra/omop_cdm.bicep:27]: Same comment, I don't think it should be secure param

File: infra/omop_cdm.bicep [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+@secure()
+@description('Admin password for the webapi database')
+param postgresWebapiAdminPassword string
+
+@secure()
+@description('Password for the cdm user')
+param postgresOMOPCDMpassword string
+
+@secure()
+@description('The name of the keyvault')
```
Comment (anatbal) [inline infra/omop_cdm.bicep:42]: Same comment

File: infra/scripts/create_omop_cdm.sh [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+pg_cdm_password="${POSTGRES_OMOP_CDM_PASSWORD}${POSTGRES_CDM_USERNAME}"
+export CDM_MD5="'md5$(echo -n $pg_cdm_password | md5sum | awk '{ print $1 }')'"
+
+printf 'Creating omp cdm schemas and user\n'
+create_omop_schemas_script=$(envsubst < create_omop_schemas.sql)
+echo "$create_omop_schemas_script" | psql "$OMOP_CONNECTION_STRING" -e
+
+# create OMOP CDM (+ Vocabulary) tables
+printf 'Creating OMOP CDM tables\n'
+sed -i  s/@cdmDatabaseSchema/${POSTGRES_OMOP_CDM_SCHEMA_NAME}/g OMOPCDM_postgresql_5.4_ddl.sql OMOPCDM_postgresql_5.4_constraints.sql OMOPCDM_postgresql_5.4_primary_keys.sql OMOPCDM_postgresql_5.4_indices.sql
```
Comment (anatbal) [inline infra/scripts/create_omop_cdm.sh:17]: why not using envsubst here as well?

File: infra/omop_cdm.bicep (hunk 7 above)
Comment (guybartal) [reply infra/omop_cdm.bicep:17]: correct, removed

File: infra/omop_cdm.bicep (hunk 5 above)
Comment (guybartal) [reply infra/omop_cdm.bicep:114]: it should run only once, removed.

**Review by anatbal — APPROVED** [review]
LGTM!

File: infra/omop_cdm.bicep (hunk 1 above)
Comment (guybartal) [reply infra/omop_cdm.bicep:68]: Not sure, I think logically those secrets should stay here in this module,
there are also advantages for having this way, 
for instance, if we want to support deployment of another OMOP CDM DB separately and independently with another button? keeping it this way will allow it to be autonomous. WDYT?

File: infra/omop_cdm.bicep (hunk 2 above)
Comment (guybartal) [reply infra/omop_cdm.bicep:74]: same as above

File: infra/omop_cdm.bicep (hunk 3 above)
Comment (guybartal) [reply infra/omop_cdm.bicep:93]: I think we should refactor our modules structure in the next iteration,
maybe this shared part should be in main along with some other shared components,
unless this identity should be deprovisioned with the script when it finished.

File: infra/omop_cdm.bicep (hunk 4 above)
Comment (guybartal) [reply infra/omop_cdm.bicep:184]: agree

File: infra/omop_cdm.bicep (hunk 1 above)
Comment (anatbal) [reply infra/omop_cdm.bicep:68]: I agree.

**Pull request merged**